specs/ChainSpec/Consensus.md

```diff
@@ -0,0 +1,146 @@
+
+If a block `B` at some height `h` builds on top of a block `B_prev` that has height `h-1`, and `B_prev` in turn builds on top of a block `B_prev_prev` that has height `h-2`, and all three blocks are in the same epoch, the block `B_prev_prev` is final, and cannot be reverted unless block producers with more than `2/3` cumulative stake deviate from the protocol. 
+
+TODO: talk about epoch boundaries here
+
+Block producers in the particular epoch exchange many kinds of messages. The two kinds that are relevant to the consensus are **Blocks** and **Approvals**. The approval contains five fields:
+
+```rust
+struct Approval {
+    prev_hash: BlockHash,
+    prev_height: BlockHeight,
+    target_height: BlockHeight,
+    signature: Signature,
+    account_id: AccountId
+}
+```
+
+Where `prev_hash` and `prev_height` are the hash and the height of the block being approved, `target_height` is the specific height at which the approval can be used (an approval with a particular `target_height` can be only included in the `approvals` of a block that has `height = target_height`), `account_id` is the account of the block producer who created the approval, and `signature` is their signature on the tuple `(prev_hash, target_height)` if `target_height == prev_height+1` and the tuple `(prev_height, target_height)` otherwise.
+
+Any valid block must contain approvals from block producers whose cumulative stake exceeds 2/3 of the total stake in the epoch. For a block `B` and its previous block `B'` each approval in `B` must have `prev_height` to be equal to the height of `B'`. If (and only if) `B.height == B'.height + 1`, each approval's `prev_hash` must also be exactly the hash of `B'`. See [this section](#approval-condition) below for details on why the condition must be exactly such.
+
+On receival of the approval message the participant just stores it in the collection of approval messages.
+
+```python
+def on_approval(self, approval):
+    self.approvals.append(approval)
+```
+
+Whenever a participant receives a block, the operations relevant to the consensus include updating the `head` and initiating a timer to start sending the approvals on the block to the block producers at the consecutive `target_height`s. The timer delays depend on the height of the last final block, so that information is also persisted.
+
+```python
+def on_block(self, block):
+    header = block.header
+
+    if header.height <= self.head_height:
+        return
+
+    last_final_block = store.get_block(header.last_final_block_hash)
+
+    self.head_height = header.height
+    self.head_hash = block.hash()
+    self.largest_final_height = last_final_block.height
+
+    self.timer_height = self.head_height + 1
+    self.timer_started = time.time()
+
+    self.endorsement_pending = True
+```
+
+The timer needs to be checked periodically, and contain the following logic:
+
+```python
+def get_delay(n):
+    min(MAX_DELAY, MIN_DELAY + DELAY_STEP * (n-2))
```

Comment:
This doesn't satisfy the conditions on _T_(_x_) in the Doomslug paper, so liveness proof doesn't apply.